fix #3863 Add an mbed API that allows the init of the CAN at the bus frequency

adustm opened the pull request:
## Description
If the bus is using a much higher frequency than the default value chosen in the init function, the device cannot synchronize and is blocked (see #3863  for details). The device needs an IDLE frame on the CAN bus to synchronize, but if the device is not set with the correct frequency, the IDLE frame may never be detected.
for instance: initialize CAN at 100kbps. It needs an idle frame on the CAN bus to synchronize (ie: 110us to have 11 consecutive recessive bits). If the bus is at 500kbs and highly busy due to another device, 110us of consecutive recessive bits does not occur. I have only 58us in my use case.
Adding the frequency parameter to the CAN class (same as done in serial api for baud rates) allows us to initialize and synchronize the CAN  device even if the bus is often busy.

## Status
**READY**

## Steps to test or reproduce
Have a 1st device with high bus occupation with the following code:
```
#include "mbed.h"
DigitalOut led2(LED2);
Serial pc(USBTX,USBRX);

int main()
{
    int mycnt=0;
    pc.baud(115200);
    pc.printf("main()\r\n");
    CAN can1(PA_11, PA_12); // only for NUCLEO_F446RE
    pc.printf("created can\r\n");
    can1.frequency(500000);
    pc.printf("set frequency \r\n");

    unsigned int id = 0x539;
    char i =0xC7;
    CANMessage tx_msg(id, &i, sizeof(i));
    bool sent = false;

    while(1) {
        if (can1.write(tx_msg)) {
            mycnt++;
            printf("Sent %u: %d\n", id, i);
            sent = true;
        }
        wait_ms(10);
        led2 = !led2;
    }
}

```
Have the current device trying to initialize. Without the PR, NUCLEO_F767ZI cannot exit from the init function (because HW cannot syncrhonize). With the PR, it can synchronize if the init function is modifying the frequency (we can see the printf "created can") with the following code:
```
#include "mbed.h"
DigitalOut led2(LED2);
Serial pc(USBTX,USBRX);

int main()
{
    pc.baud(115200);
    pc.printf("main()\r\n");
    CAN can2(PD_0, PD_1); //without the PR
    CAN can2(PD_0, PD_1, 500000); //with the PR
    pc.printf("created can\r\n");
    can2_ptr = &can2;
    //can2.frequency(500000); // without the PR
    pc.printf("set frequency \r\n");
    CANMessage msg;

    while(1) {
        if(can2.read(msg)) {
            pc.printf("<-*******Message received:id=%x,data=",msg.id);
            for(int i=0; i<8; i++) {
                pc.printf("%x,",msg.data[i] );
            }
            pc.printf("\n\r");
            led2 = !led2;
        }
    }
}

```

cc @bperry730 @ohagendorf

tommikas commented:
Getting a linking error when building with armcc at least for NUMAKER_PFM_NUC472 and UBLOX_EVK_ODIN_W2 targets.

```
Error: L6218E: Undefined symbol can_init_freq (referred from CAN.o).
Finished: 0 information, 0 warning and 1 error messages.
```

0xc0170 commented:
Thanks @tommikas . 

@adustm Adding new HAL function introduces a requirement for all targets that support CAN to have this function implemented. Therefore to make this addition, we should implement it for all targets that support CAN. There are not that many. We can create a devel branch here for this.

ohagendorf commented:
Unfortunately, I think the solution will not solve the problem in every case.

We found the problem the first time in a test application, one sending mcu and one receiving mcu. Each second a single CAN message was send with a CAN bus frequency of 1MHz. We just call the constructor without a following call of can.frequency so the default 1MHz were used. From time to time the receiving mcu didn’t start – it stopped in the constructor in can_init during the execution of can_frequency(obj, 100000) in the second while ((can->MSR & CAN_MSR_INAK) == CAN_MSR_INAK). 
The suggested PR will not solve this problem, I think.

The other application was an automotive RADAR sensor connected over CAN with an STM Nucleo. The sensor starts with an init message with a maximal message rate until it gets its config from another CAN node. In this scenario the bus will never be idle long enough to set the frequency to the requested 500kHz.

The same functionality realised only with Cube HAL function works without problems since a couple of days. We just used a CAN example from an eval board from the Cube library and adapted it to our hardware.

adustm commented:
Thanks @0xc0170  and @tommikas for your comments. I agree for the separate development branch if needed. At the moment, it seems we still have another issue. I will need to fix it before we go for the devel branch.

Hello @ohagendorf 

> The same functionality realised only with Cube HAL function works without problems since a couple of days. We just used a CAN example from an eval board from the Cube library and adapted it to our hardware.

Did you have a chance to check the differences between mbed and Cube settings wrt CAN HW registers or clocks or RCC registers ?

ohagendorf commented:
Hello @adustm  
I have the time (and the plan) to do that during the next days. I have both projects are available.

adustm commented:
Hi @ohagendorf 
Could you provide me the values that you write within your Cube sw in     
    CanHandle.Init.SJW = ;
    CanHandle.Init.BS1 = ;
    CanHandle.Init.BS2 = ;
    CanHandle.Init.Prescaler = ;

I've seen differences between Cube code and mbed-os code: the above parameters + a timeout exit.
Is it possible that you comment this timeout and check if it still works (or check if the Init function runs out with the HAL_TIMEOUT error ?) ?

Kind regards

theotherjimmy commented:
bump @adustm @ohagendorf What's the status on this?

adustm commented:
Hello @theotherjimmy 
I've been confirmed that we need to be able to initialize the CAN at the can bus frequency.
The new interface is mandatory. I am working on adding it on every CAN devices of mbed-os.

I don't have any news on the fact that it does not solve the issue from @ohagendorf but I am thinking about handling this in a separate PR (at least by detecting a timeout and by sending an error code)
Cheers

adustm commented:
Hello, I think I've added every can_init_freq for every CAN platforms.
I've compiled every boards with GCC_ARM
I've linked every mbed-5 boards with GCC_ARM and the following mail.cpp.
I hope it fits the need.
```
#include "mbed.h"

Ticker ticker;
DigitalOut led1(LED1);
DigitalOut led2(LED2);
CAN* can2_ptr;//(PD_0, PD_1);

Serial pc(USBTX,USBRX);

int main()
{
    pc.printf("main()\r\n");
  #if defined(TARGET_LPC1549)
  // LPC1549 support only single CAN channel
  CAN can2(D2, D3, 500000);
  #elif defined(TARGET_B96B_F446VE)
  // B96B_F446VE support only single CAN channel
  CAN can2(PD_0, PD_1, 500000);
  #elif defined(TARGET_NUCLEO_F091RC) || defined(TARGET_NUCLEO_F072RB) || \
        defined(TARGET_NUCLEO_F042K6) || defined(TARGET_NUCLEO_F334R8) || \
        defined(TARGET_NUCLEO_F303RE) || defined(TARGET_NUCLEO_F303K8) || \
        defined(TARGET_NUCLEO_F302R8) || defined(TARGET_NUCLEO_F446RE) || \
        defined(TARGET_DISCO_F429ZI)  || defined(TARGET_NUCLEO_F103RB) || \
        defined(TARGET_NUCLEO_F746ZG) || defined(TARGET_NUCLEO_L476RG) || \
        defined(TARGET_NUCLEO_F412ZG) || \
        defined(TARGET_NUCLEO_L432KC) || defined(TARGET_DISCO_F303VC)
  CAN can2(PA_11, PA_12, 500000);
  #elif defined(TARGET_DISCO_F469NI) ||defined(TARGET_DISCO_F746NG) || \
        defined(TARGET_NUCLEO_F207ZG)
  CAN can2(PB_8, PB_9, 500000);
  #elif defined(TARGET_NUMAKER_PFM_NUC472)
  CAN can2(PA_0, PA_1, 500000);
  #elif defined(TARGET_NUMAKER_PFM_M453)
  CAN can2(PA_13, PA_12, 500000);
  #elif defined(TARGET_UBLOX_C027)
  CAN can2(CANRD, CANTD, 500000);
  #elif defined(TARGET_RZ_A1H) || defined(TARGET_VK_RZ_A1H)
  CAN can2(P7_8, P7_9, 500000);
  #else
  CAN can2(p9, p10, 500000);
  #endif
    pc.printf("created can\r\n");
    can2_ptr = &can2;
    CANMessage msg;

    while(1) {
        if(can2.read(msg)) {
            pc.printf("<-*******Message received:id=%x,data=",msg.id);
            for(int i=0; i<8; i++) {
                pc.printf("%x,",msg.data[i] );
            }
            pc.printf("\n\r");
            led2 = !led2;
        }
    }
}

```
Kind regards
Armelle

0xc0170 commented:
@cyliangtw @TomoYamanaka @mmahadevan108 please review, as this PR touches your platforms as well. What do you think about CAN init freq proposal?

cyliangtw commented:
This PR is fine to our platforms, it could support backward compatibility.

mmahadevan108 commented:
please update can_api.c file in TARGET_LPC408X/TARGET_LPC4088_DM

adustm commented:
Thank you all for your remarks

@TomoYamanaka san,
I have added can_frequency in the can_init_freq function. Please be aware that I cannot test this modification (I don't have a platform )

Hello @mmahadevan108 
I have added the modification in LPC4088_DM can_api.c 

+ I have rebased the pull request

Kind regards
Armelle

mmahadevan108 commented:
@adustm. Thank you, changes look fine to me.

adustm commented:
Hi @0xc0170 
Could I have details about the Cam CI uvisor Build & Test failures, please ?

adbridge commented:
restart uvisor

screamerbg commented:
@0xc0170 @adbridge Can we retest this or provide info on what's failing?

adbridge commented:
It's failing due to periodic issues with the Uvisor CI. I will restart it again!

adbridge commented:
restart uvisor

adbridge commented:
@TomoYamanaka  @cyliangtw Could you please confirm you are also happy with the updates and sign off the review so that we can get this merged in ? Thanks

cyliangtw commented:
Thanks of this modification. The change is fine to me after review code.

theotherjimmy commented:
@adustm Could you include a test for this new API in this PR?

0xc0170 commented on features/unsupported/tests/mbed/can_api/main.cpp:
```diff
@@ -0,0 +1,85 @@
+#include "mbed.h"
```
please add license header to this file

adustm replied:
OK I will.
(by the way there are no license headers anywhere in unsupported/tests/mbed directory)

0xc0170 commented on features/unsupported/tests/mbed/can_api/main.cpp:
```diff
@@ -0,0 +1,85 @@
+
+    CAN can_object(CAN_RD, CAN_TD, CanFrequency);
+
+#if !defined(TARGET_VK_RZ_A1H)
```
why this exclusion ?

jeromecoutant replied:
Hi
It is a copy from can_loopback test...

adustm replied:
It is inherited from test/mbed/can_loopback/main.cpp (ie MBED_A27)
We don't have the material to test it, we did not want to take the responsibility of removing a code we don't know.
Cheers

adustm commented:
Hello @0xc0170 

> Why was this test added to unsupported ( is it because can test already it is there? )?
Correct, we have put this test with other CAN tests.


A typical test output is:
```
Test::Output::Start
MBED: Instrumentation: "COM158" and disk: "E:"
HOST: Copy image onto target...
        1 file(s) copied.
HOST: Initialize serial port...
Looking for NUCLEO_F042K6 with MBEDLS
...port ready!
HOST: Reset target...
HOST: Property 'timeout' = '20'
HOST: Property 'host_test_name' = 'dev_null'
HOST: Property 'description' = 'CAN API at different frequency'
HOST: Property 'test_id' = 'MBED_A30'
HOST: Start test...
Init CAN at 10000 Hz
Init CAN at 50000 Hz
Init CAN at 100000 Hz
Init CAN at 500000 Hz
Init CAN at 1000000 Hz
{{success}}
{{end}}
Test::Output::Finish
TargetTest::NUCLEO_F042K6[E78E]::IAR::MBED_A30::CAN API [OK] in 1.41 of 20 sec
```

And some results
```
+--------+---------------------+-----------+----------+------------------+--------------------+
| Result | Target              | Toolchain | Test ID  | Test Description | Elapsed Time (sec) |
+--------+---------------------+-----------+----------+------------------+--------------------+
| OK     | NUCLEO_F446ZE[F12F] | GCC_ARM   | MBED_A30 | CAN API          |        2.5         |
| OK     | NUCLEO_F446ZE[F12F] | IAR       | MBED_A30 | CAN API          |        2.45        |
| OK     | NUCLEO_F446ZE[F12F] | ARM       | MBED_A30 | CAN API          |        2.49        |
| OK     | NUCLEO_F429ZI[F4AE] | ARM       | MBED_A30 | CAN API          |        2.46        |
| OK     | NUCLEO_F429ZI[F4AE] | GCC_ARM   | MBED_A30 | CAN API          |        1.89        |
| OK     | NUCLEO_F429ZI[F4AE] | IAR       | MBED_A30 | CAN API          |        1.91        |
| OK     | NUCLEO_F756ZG[F009] | GCC_ARM   | MBED_A30 | CAN API          |        1.41        |
| OK     | NUCLEO_F756ZG[F009] | ARM       | MBED_A30 | CAN API          |        1.36        |
| OK     | NUCLEO_F756ZG[F009] | IAR       | MBED_A30 | CAN API          |        1.4         |
| OK     | NUCLEO_L432KC[FF26] | IAR       | MBED_A30 | CAN API          |        1.35        |
| OK     | NUCLEO_L432KC[FF26] | GCC_ARM   | MBED_A30 | CAN API          |        1.43        |
| OK     | NUCLEO_L432KC[FF26] | ARM       | MBED_A30 | CAN API          |        1.44        |
| OK     | NUCLEO_F303ZE[F746] | ARM       | MBED_A30 | CAN API          |        1.4         |
| OK     | NUCLEO_F303ZE[F746] | GCC_ARM   | MBED_A30 | CAN API          |        1.41        |
| OK     | NUCLEO_F303ZE[F746] | IAR       | MBED_A30 | CAN API          |        1.37        |
| OK     | NUCLEO_L476RG[F32D] | IAR       | MBED_A30 | CAN API          |        1.35        |
| OK     | NUCLEO_L476RG[F32D] | GCC_ARM   | MBED_A30 | CAN API          |        1.4         |
| OK     | NUCLEO_L476RG[F32D] | ARM       | MBED_A30 | CAN API          |        1.4         |
| OK     | NUCLEO_F042K6[E78E] | ARM       | MBED_A30 | CAN API          |        1.38        |
| OK     | NUCLEO_F042K6[E78E] | GCC_ARM   | MBED_A30 | CAN API          |        1.37        |
| OK     | NUCLEO_F042K6[E78E] | IAR       | MBED_A30 | CAN API          |        1.41        |
+--------+---------------------+-----------+----------+------------------+--------------------+

```
(it is not the complete list of STM32 that supports CAN, but it is a good overview of it between the STM32 families)

adustm commented:
@0xc0170 do you still need me to change the test location or is it ok for you ?

0xc0170 commented:
> @0xc0170 do you still need me to change the test location or is it ok for you ?

Should be fine. Please rebase if this has all the work done, we will review it once done

adustm commented:
Thanks @0xc0170 
Rebase done: target LPC23XX can file deleted + manual merge for TARGET_NXP/TARGET_LPC408X/TARGET_LPC4088_DM/can_api.c

0xc0170 commented on drivers/CAN.h:
```diff
@@ -111,6 +111,15 @@ class CAN {
+      * @param td the td pin
+      * @param f the bus frequency in hertz
+      */
+    CAN(PinName rd, PinName td, int f);
```
Can you please name the parameter f the full name ? does not make sense to short it as it was some loop counter? Follow what is in the another API (baud(int baudrate), frequency(int hz), etc).

adustm replied:
there is already a function called frequency. If you agree, I will call it hz as in the frequency function.
Ok for you ?

ohagendorf commented:
I'm very sorry for the late answer.

I checked the latest version. With the possibility of setting the CAN bus frequency in the constructor the problem is solved. 

But using the frequency method is still not possible. 

In can_frequency two loops are checking a status bit in the CAN bus register set. Without the correct bit setting the function will never return. That is the problem with a busy CAN bus and using this function after the instantiation. With a timeout check within these loops this behaviour could be resolved. And a return value for a successful/unsuccessful frequency setting is available.

ohagendorf commented:
I checked the branch mentioned by @adustm in #3863. 
That is a behaviour I would expect: When calling the frequency method with a busy CAN bus I'm getting an error message at the serial port and a return value of 0 saying something was not successful.

adustm commented:
Thanks @ohagendorf 
I will finalize this PR for the addition of the frequency parameter, then I will create the other PR to add the timeout and error check in case the bus is too bus to initialize the Can hardware block

@0xc0170 , I will rename the f into frequency if you like.

adustm commented:
quick test on the last modification : 
```
+--------+---------------------+-----------+----------+------------------+--------------------+---------------+-------+
| Result | Target              | Toolchain | Test ID  | Test Description | Elapsed Time (sec) | Timeout (sec) | Loops |
+--------+---------------------+-----------+----------+------------------+--------------------+---------------+-------+
| OK     | NUCLEO_F756ZG[F009] | GCC_ARM   | MBED_A30 | CAN API          |        1.36        |       20      |  1/1  |
| OK     | NUCLEO_F439ZI[F4A5] | GCC_ARM   | MBED_A30 | CAN API          |        1.36        |       20      |  1/1  |
+--------+---------------------+-----------+----------+------------------+--------------------+---------------+-------+
```

0xc0170 commented:
/morph test-nightly

adustm commented:
> Cam-CI uvisor Build & Test — Failed 
?

0xc0170 commented:
/morph test-nightly

0xc0170 commented:
retest uvisor

adbridge commented:
/morph test-nightly

adbridge commented:
retest uvisor

0xc0170 commented:
/morph test-nightly

0xc0170 commented:
retest uvisor

adbridge commented:
/morph test-nightly

adbridge commented:
retest uvisor

bridadan commented:
We had a K64F that was behind on firmware updates and was showing failures. It has been updated now.

/morph test-nightly

adustm commented:
Hello @bridadan @adbridge , the K64F is in TIMEOUT-ERROR. 
I just tested an extract of it myself and here is the result:
```
+--------------+---------------+----------------------------------+--------+--------------------+-------------+
| target       | platform_name | test suite                       | result | elapsed_time (sec) | copy_method |
+--------------+---------------+----------------------------------+--------+--------------------+-------------+
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | OK     | 18.3               | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-dev_null      | OK     | 20.76              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-echo          | OK     | 29.66              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-flashiap      | OK     | 16.7               | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-generic_tests | OK     | 18.36              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | OK     | 25.67              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-rtc           | OK     | 28.12              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-stl_features  | OK     | 18.61              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-ticker        | OK     | 49.93              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-timeout       | OK     | 27.49              | shell       |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-wait_us       | OK     | 27.48              | shell       |
+--------------+---------------+----------------------------------+--------+--------------------+-------------+
mbedgt: test suite results: 11 OK
mbedgt: test case report:
+--------------+---------------+----------------------------------+-------------------------------------+--------+--------+--------+--------------------+
| target       | platform_name | test suite                       | test case                           | passed | failed | result | elapsed_time (sec) |
+--------------+---------------+----------------------------------+-------------------------------------+--------+--------+--------+--------------------+
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %e %E float formatting   | 1      | 0      | OK     | 0.06               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %f %f float formatting   | 1      | 0      | OK     | 0.06               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %g %g float formatting   | 1      | 0      | OK     | 0.06               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %i %d integer formatting | 1      | 0      | OK     | 0.07               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %u %d integer formatting | 1      | 0      | OK     | 0.07               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: %x %E integer formatting | 1      | 0      | OK     | 0.06               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: strpbrk                  | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-c_strings     | C strings: strtok                   | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-dev_null      | tests-mbed_drivers-dev_null         | 1      | 0      | OK     | 20.76              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-echo          | Echo server: x16                    | 1      | 0      | OK     | 1.99               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-echo          | Echo server: x32                    | 1      | 0      | OK     | 3.84               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-echo          | Echo server: x64                    | 1      | 0      | OK     | 7.58               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-flashiap      | FlashIAP - init                     | 1      | 0      | OK     | 0.04               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-flashiap      | FlashIAP - program                  | 1      | 0      | OK     | 0.1                |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-flashiap      | FlashIAP - program errors           | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-generic_tests | Basic                               | 1      | 0      | OK     | 0.03               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-generic_tests | Blinky                              | 1      | 0      | OK     | 0.03               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-generic_tests | C++ heap                            | 1      | 0      | OK     | 0.09               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-generic_tests | C++ stack                           | 1      | 0      | OK     | 0.15               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 1ms LowPowerTimeout                 | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 1sec LowPowerTimeout                | 1      | 0      | OK     | 1.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 1sec LowPowerTimeout from deepsleep | 1      | 0      | OK     | 1.08               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 1sec LowPowerTimeout from sleep     | 1      | 0      | OK     | 1.06               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 500us LowPowerTimeout               | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-lp_timeout    | 5sec LowPowerTimeout                | 1      | 0      | OK     | 5.04               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-rtc           | RTC strftime                        | 1      | 0      | OK     | 10.53              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-stl_features  | STL std::equal                      | 1      | 0      | OK     | 0.04               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-stl_features  | STL std::sort abs                   | 1      | 0      | OK     | 0.04               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-stl_features  | STL std::sort greater               | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-stl_features  | STL std::transform                  | 1      | 0      | OK     | 0.05               |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-ticker        | Timers: 1x ticker                   | 1      | 0      | OK     | 11.07              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-ticker        | Timers: 2x callbacks                | 1      | 0      | OK     | 11.06              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-ticker        | Timers: 2x tickers                  | 1      | 0      | OK     | 11.06              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-timeout       | Timers: toggle on/off               | 1      | 0      | OK     | 11.07              |
| K64F-GCC_ARM | K64F          | tests-mbed_drivers-wait_us       | Timers: wait_us                     | 1      | 0      | OK     | 11.17              |
+--------------+---------------+----------------------------------+-------------------------------------+--------+--------+--------+--------------------+
mbedgt: test case results: 35 OK
```

I guess there is an issue in your testbench somewhere. Could you try and solve it, please ?

adbridge commented:
/morph test-nightly

adbridge commented:
retest uvisor

adbridge commented:
@adustm I've restarted the jobs, if they fail again we may have a broken device in the CI @studavekar

studavekar commented:
Apologies have aborted few of the jobs which failed and on one of device. CI seems to unstable now with the devices.

Will re-trigger the tests after checking few jobs.

studavekar commented:
/morph test-nightly

sg- commented:
Looks good although the can_init looks useless at this point so should be considered for deprecation. Second note is that HAL setting the default frequency seems wrong, although was the was it was apparently. Shall we remove can_init in another PR and use the constructor to set a default frequency across all CAN object using ```CAN::CAN(PinName rd, PinName td)```

adustm commented:
> Shall we remove can_init in another PR and use the constructor to set a default frequency across all CAN object using CAN::CAN(PinName rd, PinName td)
@sg- for sure the old can_init is not usefull anymore
